In [2]:
%matplotlib inline
import numpy as np
import scipy as sp
import matplotlib as mpl
import matplotlib.cm as cm
import matplotlib.pyplot as plt
import pandas as pd
pd.set_option('display.width', 500)
pd.set_option('display.max_columns', 100)
pd.set_option('display.notebook_repr_html', True)
import seaborn as sns
sns.set_style("whitegrid")
sns.set_context("poster")

In [23]:
df=pd.read_csv("tempdata/30122-0001-Data.tsv", sep='\t')

In [3]:
df.head()

CASEID  DISYR  AGE  GENDER  RACE  ETHNIC  MARSTAT  EDUC  EMPLOY  DETNLF  PREG  VET  LIVARAG  PRIMINC  ARRESTS  STFIPS   CBSA  PMSA  REGION  DIVISION  SERVSETD  METHUSE  DAYWAIT  REASON  LOS  PSOURCE  DETCRIM  NOPRIOR  SUB1  ROUTE1  FREQ1  FRSTUSE1  SUB2  ROUTE2  FREQ2  FRSTUSE2  SUB3  ROUTE3  FREQ3  FRSTUSE3  NUMSUBS  IDU  ALCFLG  COKEFLG  MARFLG  HERFLG  METHFLG  OPSYNFLG  PCPFLG  HALLFLG  MTHAMFLG  AMPHFLG  STIMFLG  BENZFLG  TRNQFLG  BARBFLG  SEDHPFLG  INHFLG  OTCFLG  OTHERFLG  \
0  20060000001   2006    9       1     5       5        4     2       3      -9    -9    2        3       -9       -8      26  19820  2160       2         3         7        2        0       5   25        1       -9        1     2       1      3         3     4       2      5         3     1      -9     -9        -9        2    0       1        0       1       0        0         0       0        0         0        0        0        0        0        0         0       0       0         0   
1  20060000002   2006   12       1     5       5        4     3       1      -9    -9    2        3        1       -8      29     -9    -9       2         4         2        2       -9       4    5        7        1        0     2       1      5         3     1      -9     -9        -9     1      -9     -9        -9        1    0       1        0       0       0        0         0       0        0         0        0        0        0        0        0         0       0       0         0   
2  20060000003   2006    7       2     5       5       -9     3       4       1     2    2        3       20       -8      27     -9    -9       2         4         6        2       -9       1   35        6       -9       -9     2       1      1         2     1      -9     -9        -9     1      -9     -9        -9        1    0       1        0       0       0        0         0       0        0         0        0        0        0        0        0         0       0       0         0   
3  20060000004   2006    7       2     5       5        4     3       1      -9     2    2        2        1       -8      17  41180  7040       2         3         7        2        0       3   35        7        8        2    10       2      1         3     4       2      1        -9     1      -9     -9        -9        2    0       0        0       1       0        0         0       0        0         1        0        0        0        0        0         0       0       0         0   
4  20060000005   2006    9       1     4       5        1     2       4       6    -9    2        3       21       -8      19  19340  1960       2         4         4        2        0       1   17        7       -9        0     3       3      2         9     4       2      5         5     1      -9     -9        -9        2    0       0        1       1       0        0         0       0        0         0        0        0        0        0        0         0       0       0         0   

   ALCDRUG  DSMCRIT  PSYPROB  HLTHINS  PRIMPAY  
0        3        4        2       -9       -9  
1        1       -9        2        4        5  
2        1       -9       -9       -9       -9  
3        2        8        2        4       -9  
4        2       -9        2       -9        5

In [4]:
df.shape

(9829536, 65)

In [5]:
df.describe()

CASEID           DISYR             AGE          GENDER            RACE          ETHNIC         MARSTAT            EDUC          EMPLOY          DETNLF            PREG             VET         LIVARAG         PRIMINC         ARRESTS          STFIPS            CBSA            PMSA          REGION        DIVISION        SERVSETD         METHUSE         DAYWAIT          REASON             LOS         PSOURCE         DETCRIM         NOPRIOR            SUB1          ROUTE1  \
count  9.829536e+06  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000  9829536.000000   
mean   2.008648e+10     2008.565576        7.385672        1.316786        6.243348        4.460965       -0.730916        2.589892        2.756016       -4.659835       -5.643938        0.389218        2.225300        4.357847       -3.050121       27.941288    23554.174972     3443.690643        2.418327        4.746925        5.348798        1.416218       -0.626684        2.235459       23.072387        3.823122       -5.502129        0.234684        4.027725        1.737084   
std    1.704791e+07        1.701035        2.539206        0.550357        5.244070        1.910116        4.683082        1.903108        1.993942        6.401648        5.051948        3.842441        1.891297       12.773499        4.046700       14.461250    16196.133161     2915.210062        1.171805        2.805823        2.062796        2.353995       31.729369        1.608214       13.392682        3.164659        5.910499        3.722664        2.899761        2.116117   
min    2.006000e+10     2006.000000        2.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000        1.000000       -9.000000       -9.000000        0.000000        0.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000       -9.000000   
25%    2.007095e+10     2007.000000        5.000000        1.000000        4.000000        5.000000        1.000000        2.000000        2.000000       -9.000000       -9.000000        2.000000        2.000000       -9.000000       -8.000000       17.000000    10780.000000      220.000000        1.000000        2.000000        4.000000        2.000000       -9.000000        1.000000        7.000000        1.000000       -9.000000        0.000000        2.000000        1.000000   
50%    2.009012e+10     2009.000000        7.000000        1.000000        5.000000        5.000000        1.000000        3.000000        3.000000       -9.000000       -9.000000        2.000000        3.000000        1.000000        0.000000       29.000000    26420.000000     3120.000000        2.000000        4.000000        6.000000        2.000000        0.000000        2.000000       31.000000        3.000000       -9.000000        1.000000        3.000000        2.000000   
75%    2.010096e+10     2010.000000        9.000000        2.000000        5.000000        5.000000        2.000000        3.000000        4.000000        3.000000        2.000000        2.000000        3.000000       20.000000        0.000000       36.000000    36140.000000     5640.000000        4.000000        8.000000        7.000000        2.000000        0.000000        3.000000       34.000000        7.000000        1.000000        2.000000        5.000000        2.000000   
max    2.011173e+10     2011.000000       12.000000        2.000000       23.000000        6.000000        4.000000  

In [6]:
small_df=df.copy()
for i in small_df.columns:
    small_df=small_df[small_df[i]>=0]    

In [7]:
small_df.shape

(2287, 65)

After dropping all rows with a missing value, we end up with only 2,287 samples. The sample is quite biased as all of the samples come from 2008 and are only female. As this method is not suitable for our purposes, we must deal with the missing values in a different way.

In [8]:
small_df.head(5)

CASEID  DISYR  AGE  GENDER  RACE  ETHNIC  MARSTAT  EDUC  EMPLOY  DETNLF  PREG  VET  LIVARAG  PRIMINC  ARRESTS  STFIPS   CBSA  PMSA  REGION  DIVISION  SERVSETD  METHUSE  DAYWAIT  REASON  LOS  PSOURCE  DETCRIM  NOPRIOR  SUB1  ROUTE1  FREQ1  FRSTUSE1  SUB2  ROUTE2  FREQ2  FRSTUSE2  SUB3  ROUTE3  FREQ3  FRSTUSE3  NUMSUBS  IDU  ALCFLG  COKEFLG  MARFLG  HERFLG  METHFLG  OPSYNFLG  PCPFLG  HALLFLG  MTHAMFLG  AMPHFLG  STIMFLG  BENZFLG  TRNQFLG  BARBFLG  SEDHPFLG  INHFLG  OTCFLG  OTHERFLG  \
3116475  20080000633   2008    3       2     5       5        1     2       4       2     2    2        2       21        0      17  14060  1040       2         3         4        2       72       1   33        7        3        2     4       2      4         2     2       1      2         2    12       2      5         2        3    0       1        0       1       0        0         0       0        0         0        0        1        0        0        0         0       0       0         0   
3133109  20080017267   2008    7       2     2       5        1     3       4       1     2    2        2       21        0      17  16980  1600       2         3         7        2       30       2   34        7        3        3     2       1      1         3     3       3      1         4     4       2      1         4        3    0       1        1       1       0        0         0       0        0         0        0        0        0        0        0         0       0       0         0   
3134727  20080018885   2008    3       2     5       5        1     1       4       6     2    2        2        1        0      17  14060  1040       2         3         4        2        0       2   13        7        3        0     4       2      5         1     2       1      2         2    12       2      5         2        3    0       1        0       1       0        0         0       0        0         0        0        1        0        0        0         0       0       0         0   
3141417  20080025575   2008    3       2     5       5        1     2       4       2     2    2        2        3        0      17  41180  7040       2         3         7        2       10       2   33        7        5        0     2       1      2         3     4       2      3         3    12       2      5         3        3    0       1        0       1       0        0         0       0        0         0        0        1        0        0        0         0       0       0         0   
3141427  20080025585   2008    8       2     5       5        1     3       4       1     2    2        2       21        0      17  16980  1600       2         3         7        2       11       4   34        7        3        1     2       1      2         3     3       3      2         7     4       2      2         3        3    0       1        1       1       0        0         0       0        0         0        0        0        0        0        0         0       0       0         0   

         ALCDRUG  DSMCRIT  PSYPROB  HLTHINS  PRIMPAY  
3116475        3        7        2        2        4  
3133109        3        4        2        2        4  
3134727        3        7        2        2        4  
3141417        3        7        2        1        1  
3141427        3        9        1        2        4

In [9]:
df.MARSTAT.value_counts()

 1    4593570
-9    2302435
 4    1308432
 2    1165247
 3     459852
dtype: int64

In [10]:
df[df.MARSTAT==-9].DISYR.value_counts()

2008    409444
2009    406584
2007    405468
2006    363861
2010    362695
2011    354383
dtype: int64

In [11]:
df[df.PREG==-9].GENDER.value_counts()

 1    6630689
 2     180443
-9       7708
dtype: int64

In [12]:
len(df.CBSA.unique())

273

- `CASEID: CASE IDENTIFICATION NUMBER`
- `DISYR: YEAR OF DISCHARGE`
- `AGE: AGE (RECODED)`
- `GENDER: SEX`
- `RACE: RACE`
- `ETHNIC: ETHNICITY (HISPANIC ORIGIN)`
- `MARSTAT: MARITAL STATUS`
- `EDUC: EDUCATION`
- `EMPLOY: EMPLOYMENT STATUS`
- `DETNLF: DETAILED 'NOT IN LABOR FORCE' CATEGORY`
- `PREG: PREGNANT AT TIME OF ADMISSION`
- `VET: VETERAN STATUS`
- `LIVARAG: LIVING ARRANGEMENTS`
- `PRIMINC: SOURCE OF INCOME/SUPPORT`
- `ARRESTS: NUMBER OF ARRESTS IN 30 DAYS PRIOR TO ADMISSION`
- `STFIPS: CENSUS STATE FIPS CODE`
- `CBSA: FIPS 2000 CBSA CODE`
- `PMSA: FIPS 1990 MSA CODE`
- `REGION: CENSUS REGION`
- `DIVISION: CENSUS DIVISION`
- `SERVSETD: SERVICE SETTING AT DISCHARGE`
- `METHUSE: MEDICATION-ASSISTED OPIOID THERAPY`
- `DAYWAIT: DAYS WAITING TO ENTER TREATMENT`
- `REASON: REASON FOR DISCHARGE`
- `LOS: LENGTH OF STAY`
- `PSOURCE: PRINCIPAL SOURCE OF REFERRAL`
- `DETCRIM: DETAILED CRIMINAL JUSTICE REFERRAL`
- `NOPRIOR: NUMBER OF PRIOR TREATMENT EPISODES`
- `SUB1: SUBSTANCE PROBLEM CODE (PRIMARY)`
- `ROUTE1: USUAL ROUTE OF ADMINISTRATION (PRIMARY)`
- `FREQ1: FREQUENCY OF USE (PRIMARY)`
- `FRSTUSE1: AGE AT FIRST USE (PRIMARY)`
- `SUB2: SUBSTANCE PROBLEM CODE (SECONDARY)`
- `ROUTE2: USUAL ROUTE OF ADMINISTRATION (SECONDARY)`
- `FREQ2: FREQUENCY OF USE (SECONDARY)`
- `FRSTUSE2: AGE AT FIRST USE (SECONDARY)`
- `SUB3: SUBSTANCE PROBLEM CODE (TERTIARY)`
- `ROUTE3: USUAL ROUTE OF ADMINISTRATION (TERTIARY)`
- `FREQ3: FREQUENCY OF USE (TERTIARY)`
- `FRSTUSE3: AGE AT FIRST USE (TERTIARY)`
- `NUMSUBS: NUMBER OF SUBSTANCES REPORTED AT ADM.`
- `IDU: CURRENT IV DRUG USE REPORTED AT ADM.`
- `ALCFLG` and 17 other Indicator variables for individual drug types
- `ALCDRUG: SUBSTANCE ABUSE TYPE`
- `DSMCRIT: DSM DIAGNOSIS`
- `PSYPROB: PSYCHIATRIC PROBLEM IN ADDITION TO ALCOHOL/DRUG PROBLEM`
- `HLTHINS: HEALTH INSURANCE`
- `PRIMPAY: EXPECTED/ACTUAL PRIMARY SOURCE OF PAYMENT`

Find more details about class designation for each variable at the following link:
http://wwwdasis.samhsa.gov/dasis2/TEDS%20Pubs/TEDS/Discharges/TED-D-Concatenated%202006%20to%202011/30122-0001-Codebook.pdf


In [13]:
df.shape[0]

9829536

In [24]:
drop_col=['ARRESTS','DAYWAIT','LOS','SUB3','ROUTE3','FREQ3','FRSTUSE3','ALCFLG','COKEFLG','MARFLG','HERFLG',
      'METHFLG','OPSYNFLG','PCPFLG','HALLFLG','MTHAMFLG','AMPHFLG','STIMFLG','BENZFLG','TRNQFLG','BARBFLG','SEDHPFLG',
      'INHFLG','OTCFLG','OTHERFLG']
df.drop(drop_col, axis=1, inplace=True)

## BREAKDOWN OF RESPONSE VARIABLE:##

REASON: REASON FOR DISCHARGE
Indicates the outcome of treatment or the reason for transfer or discontinuance of treatment.

- **TREATMENT COMPLETED (1)**: All parts of the treatment plan or program were completed.
- **LEFT AGAINST PROFESSIONAL ADVICE (2)**: Client chose not to complete program, with or without specific advice to continue treatment. Includes clients who "drop out" of treatment for unknown reason and clients who have not received treatment for some time and are discharged for "administrative" reasons.
- **TERMINATED BY FACILITY (3)**: Treatment terminated by action of facility, generally because of client non-compliance or violation of rules, laws, or procedures (not because client dropped out of treatment, client incarcerated, or other client motivated reason).
- **TRANSFERRED TO ANOTHER SUBSTANCE ABUSE TREATMENT PROGRAM OR FACILITY (4)**: Client was transferred to another substance abuse treatment program, provider or facility within an episode of treatment. 
- **INCARCERATED (5)**: This code is to be used for all clients whose course of treatment is terminated because the client has been incarcerated. Includes jail, prison, and house confinement.
- **DEATH (6)**
- **OTHER (7)**: Moved, illness, hospitalization, or other reason somewhat out of client's control.
- **UNKNOWN (8)**: Client status at discharge is not known because for example, discharge record is lost or incomplete.

>Because both treatment completion and transfer to further treatment represent positive conclusions
to a treatment episode or component of a treatment episode, their rates are combined in some of
the analyses. (PAGE 5)

In [25]:
df=df[(df.REASON>0) & (df.REASON<6)]

In [26]:
df['RESPONSE']=[1 if (i==1) | (i==4) else 0 for i in df.REASON]

**In the original dataset, some features are dependent on others. Since these "missing" values are not missing completely at random, deleting them would introduce some bias into the dataset. For this reason, a new category is created for values that are not applicable.**

In [27]:
df.reset_index(drop=True, inplace=True)
df['PREG']=[2 if df.GENDER[i]==1 else val for i,val in enumerate(df.PREG)]
#for clients in the labor force, create a new category for "in labor force" (7) in the DETNLF instead of categorizing
#as missing
df['DETNLF']=[7 if df.EMPLOY[i]!=4 else val for i,val in enumerate(df.DETNLF)]

df['ROUTE1']=[21 if df.SUB1[i]==1 else val for i,val in enumerate(df.ROUTE1)]
df['FREQ1']=[6 if df.SUB1[i]==1 else val for i,val in enumerate(df.FREQ1)]
df['FRSTUSE1']=[13 if df.SUB1[i]==1 else val for i,val in enumerate(df.FRSTUSE1)]
df['ROUTE2']=[21 if df.SUB2[i]==1 else val for i,val in enumerate(df.ROUTE2)]
df['FREQ2']=[6 if df.SUB2[i]==1 else val for i,val in enumerate(df.FREQ2)]
df['FRSTUSE2']=[13 if df.SUB2[i]==1 else val for i,val in enumerate(df.FRSTUSE2)]

df['DETCRIM']=[9 if df.PSOURCE[i]!=7 else val for i,val in enumerate(df.DETCRIM)]
df['IDU']=[2 if i==-8 else i for i in df.IDU]

In [28]:
missing_dict={}
for i in df.columns:
    missing_dict[i]=df[df[i]<0].shape[0]/float(df.shape[0])
missing_dict

{'AGE': 0.0,
 'ALCDRUG': 0.0,
 'CASEID': 0.0,
 'CBSA': 0.23076635106640817,
 'DETCRIM': 0.08759130840481072,
 'DETNLF': 0.08564774897478235,
 'DISYR': 0.0,
 'DIVISION': 0.0,
 'DSMCRIT': 0.649064295596596,
 'EDUC': 0.019319628336616705,
 'EMPLOY': 0.020066307143492605,
 'ETHNIC': 0.01260747963253635,
 'FREQ1': 0.009194336687058963,
 'FREQ2': 0.01619403364798693,
 'FRSTUSE1': 0.010864009898630903,
 'FRSTUSE2': 0.02430939202745833,
 'GENDER': 0.0008174632262570178,
 'HLTHINS': 0.5650418988899433,
 'IDU': 0.0,
 'LIVARAG': 0.023429535240044704,
 'MARSTAT': 0.2386609653295117,
 'METHUSE': 0.04718577607531806,
 'NOPRIOR': 0.11978477233503923,
 'NUMSUBS': 0.0,
 'PMSA': 0.23076635106640817,
 'PREG': 0.01932963282142285,
 'PRIMINC': 0.3745905019142452,
 'PRIMPAY': 0.6963926086866252,
 'PSOURCE': 0.01669716241614117,
 'PSYPROB': 0.27950518893899856,
 'RACE': 0.011934059472466743,
 'REASON': 0.0,
 'REGION': 0.0,
 'RESPONSE': 0.0,
 'ROUTE1': 0.008999518171102724,
 'ROUTE2': 0.01582053288189082,
 'S

In [30]:
df.DETCRIM.value_counts()/df.DETCRIM.count()

 9    0.640569
 3    0.110660
-9    0.087591
 1    0.067524
 8    0.054199
 7    0.024807
 5    0.008251
 6    0.006397
dtype: float64

In [31]:
df.to_csv('tempdata/new_data.csv')

In [3]:
df=pd.read_csv('tempdata/new_data.csv')

TESTING TESTING

In [55]:
drop_nan=[]
for var, val in missing_dict.iteritems():
    if (val<0.02) & (val>0):
        drop_nan.append(var)
drop_nan

['FRSTUSE1',
 'ETHNIC',
 'EDUC',
 'PSOURCE',
 'SUB2',
 'FREQ2',
 'FREQ1',
 'RACE',
 'SERVSETD',
 'ROUTE2',
 'ROUTE1',
 'SUB1',
 'GENDER',
 'PREG']

In [58]:
for i in drop_nan:
    df=df[df[i]>=0]   

In [59]:
df.shape

(8439258, 42)

In [39]:
make_mode=[]
for var, val in missing_dict.iteritems():
    if (val<0.05) & (val>0.02):
        make_mode.append(var)
make_mode

['FRSTUSE2', 'EMPLOY', 'LIVARAG', 'METHUSE']

In [132]:
def impute_missing(df, X_cols, imp_cols, n_train):
    from sklearn.ensemble import RandomForestClassifier
    input_df=df.copy()
    for i in imp_cols:
        mask=input_df[i]<0
        no_missing=input_df[~mask]
        reduced=no_missing.sample(n=n_train, random_state=50)
        
        xtrain=reduced[X_cols]
        ytrain=reduced[i]
        xtest=input_df[X_cols][mask]
        
        #account for unbalanced classes
        clf=RandomForestClassifier(n_estimators=50, random_state=50, oob_score=True, class_weight='subsample')
        clf.fit(xtrain, ytrain)
        y_hat=clf.predict(xtest)
        input_df[i][mask]=y_hat
        
        print "FINISHED: "+ i + " with out-of-bag score= %f0.3" %clf.oob_score_
    return input_df

In [ ]:
test=impute_missing(df, ['AGE', 'ALCDRUG','DIVISION','NUMSUBS'], ['MARSTAT','VET'], 10000)

FINISHED: MARSTAT with out-of-bag score= 0.4589000.3
FINISHED: VET with out-of-bag score= 0.7377000.3


In [121]:
test.VET.value_counts()

2    8954790
1     341041
dtype: int64

In [122]:
df.VET.value_counts()

 2    7635938
-9    1326748
 1     333145
dtype: int64

In [118]:
test.MARSTAT.value_counts()

1    5992590
4    1694728
2    1172025
3     436488
dtype: int64

In [111]:
len(test.MARSTAT)

9295831

In [119]:
df.MARSTAT.value_counts()

 1    4319603
-9    2218552
 4    1226099
 2    1095632
 3     435945
dtype: int64

In [113]:
len(df.MARSTAT)

9295831

In [ ]:
#MAKE SURE TO DO THIS SEPARATELY ON THE TEST / VAL / TRAIN SETS
from sklearn.preprocessing import Imputer
imp = Imputer(missing_values=-9, strategy='most_frequent', axis=0)
df[make_mode] = imp.fit_transform(df[make_mode])

In [ ]:
from scipy.stats.stats import pearsonr
correlations=[]
dftousetrain=dftouse[mask]
for col in lcols:
    r=pearsonr(dftousetrain[col], dftousetrain['RESP'])[0]
    correlations.append(dict(feature=col,corr=r, abscorr=np.abs(r)))

bpdf=pd.DataFrame(correlations).sort('abscorr', ascending=False)
bpdf.set_index(['feature'], inplace=True)
bpdf.head(25)